### PR TITLE
[DistillationTrainer refactor] Loss consumes `completion_mask`

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 19 (Group B) — stacked on PR 18. Part of #6449.*

The loss now masks on `completion_mask` instead of deriving `labels != -100`; `labels` is retained for metrics only. Second step of the layout migration (add → switch → delete).

- `_reduce_divergence_loss` takes `completion_mask` (not `labels`). `generalized_jsd_loss` keeps its GKD-pinned `labels` API and converts internally, so the cross-trainer parity test is untouched.
- Main `compute_loss` computes per-token JSD (`reduction="none"`) then reduces by `inputs["completion_mask"]`; the Liger path and the server's two call sites switch too.
- The collator also emits `completion_mask` (all-zeros, prompt-only) so the eval path — which uses the collator batch without generation — has the key. This completes item 18's emission, which only covered the generation path.

Verified: 52 tests pass (distillation + server); ruff clean.

---

> [!NOTE]
> **Medium Risk**
> Touches core loss reduction and gradient-accumulation denominators; behavior should be equivalent if masks match `labels != -100`, but any mask/label drift would change training objectives.
> 
> **Overview**
> **Distillation loss masking** now uses an explicit `completion_mask` tensor instead of inferring valid tokens from `labels != -100`. `labels` stays on batches for metrics and GKD-compatible APIs.
> 
> `_reduce_divergence_loss` takes `completion_mask`; `generalized_jsd_loss` still accepts `labels` and converts internally so cross-trainer parity tests are unchanged. Main `compute_loss` gets per-token JSD with `reduction="none"`, then reduces via `inputs["completion_mask"]` (prompt-sliced). The Liger and server distillation paths pass the same mask into reduction.
> 
> `_DistillationCollator` adds `completion_mask` on the prompt-only eval path (all zeros before generation). Tests that record `num_items_in_batch` and synthetic loss batches were updated for the new signature and batch keys.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 80d20fb8ac938ce2ac9eb3e6db46ee2dee36ab1f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/server_distillation/server_distillation_trainer.py [hunk 1]
```diff
@@ -191,7 +191,7 @@ def _compute_sparse_top_1_divergence_loss(
 
         jsd = _jsd_divergence(student_sparse_log_probs, teacher_sparse_log_probs, self.beta, support_mask)
         return self._reduce_divergence_loss(
-            jsd, labels=labels, reduction="batchmean", num_items_in_batch=num_items_in_batch
+            jsd, completion_mask=labels != -100, reduction="batchmean", num_items_in_batch=num_items_in_batch
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/server_distillation/server_distillation_trainer.py:194]: The server masks the identical positions: `trimmed_labels != -100` equals the teacher-window-trimmed completion_mask (they're equal at emission, and both are sliced the same way), so there's no behavioral divergence, only the intentional, documented window-limiting. Threading the completion_mask key through the server helpers is the job of item 21 ("switch all consumers to the new keys"); item 19 keeps the equivalent labels-derived mask.

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -976,19 +980,21 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
             attention_mask=inputs["attention_mask"],
         )
         prompt_length = self._compute_prompt_length(inputs)
-        labels = inputs["labels"][:, prompt_length:]
+        completion_mask = inputs["completion_mask"][:, prompt_length:]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:983]: True for a custom old-schema collator, but the default collator now emits completion_mask, this is experimental API mid-migration, and the collator becomes identity at item 26; a labels-fallback would be defensive code removed two steps later. Leaving as-is per the repo's simplicity rule.

File: trl/experimental/distillation/distillation_trainer.py [hunk 3]
```diff
@@ -976,19 +980,21 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
 
         # Local teacher: exact full-vocabulary (optionally top-k) generalized JSD/KL loss.
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py]: Nit: Should we remove this comment?

**Review by cmpatino — APPROVED** [review]
LGTM! Would be nice to remove the stale comment but it's nothing blocking the PR

File: trl/experimental/distillation/distillation_trainer.py (hunk 3 above)
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py]: Thanks, done, removed; it was stale anyway (top-k support was dropped at #6456).
